core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -312,13 +361,38 @@ private RexNode simplifyNot(RexCall call) {
   private RexNode simplifyIs(RexCall call) {
     final SqlKind kind = call.getKind();
     final RexNode a = call.getOperands().get(0);
+
+    final RexNode pred = simplifyIsPredicate(kind, a);
+    if (pred != null) {
+      return pred;
+    }
+
     final RexNode simplified = simplifyIs2(kind, a);
     if (simplified != null) {
       return simplified;
     }
     return call;
   }
 
+  private RexNode simplifyIsPredicate(SqlKind kind, RexNode a) {
+    if (!RexUtil.isReferenceOrAccess(a, true)) {
+      return null;
+    }
+
+    for (RexNode p : predicates.pulledUpPredicates) {
+      IsPredicate pred = IsPredicate.of(p);
+      if (pred == null || !a.toString().equals(pred.ref.toString())) {
+        continue;
+      }
+      if (kind == pred.kind) {
+        return rexBuilder.makeLiteral(true);
+      } else {
+        return rexBuilder.makeLiteral(false);
```

Comment:
This is wrong.

Counter-example is `OR(IS NOT NULL($10), IS NOT FALSE($10))` where `$10` is a nullable boolean.

It is trying to simplify `is not false($10)` with assumption of `NOT(is not null($10))` which is `is null($10)`.
It sees that it has assumption for `IS NULL($10)` and its kind does not match to `IS NOT FALSE`, thus it simplifies the predicate to FALSE which is wrong.
`IS NOT FALSE(null)` yields `true`.

Reply:
@vlsi , was this issue ever addressed? Going through this logic to fix CALCITE-2639 .

Reply:
I've no idea.